File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations
from math import sqrt


def is_prime(n):
    if n < 2:
        return False
    
    for i in range(2, int(sqrt(n)) + 1):
        if n % i == 0:
            return False
        
    return True


def solution(numbers):
    answer = 0
    n = len(numbers)
    
    papers = list(numbers)
    
    made_nums = set()
    
    for i in range(1, n + 1):
        for permu in permutations(papers, i):
            made_nums.add(int(''.join(permu)))
    
    for made_num in made_nums:
        if is_prime(made_num):
            answer += 1
            
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.kt
import kotlin.math.sqrt

class Solution {
    fun makePermutation(char: List<Char>): Set<Int> {
        val result = mutableSetOf<Int>()
        
        fun permute(prefix: String, remain: String) {
            if (prefix.isNotEmpty()) result.add(prefix.toInt())
            
            for (i in remain.indices) {
                permute(prefix + remain[i], remain.removeRange(i, i + 1))
            }
        }
        
        permute("", char.joinToString(""))
        
        return result
    }
    
    fun isPrime(num: Int): Boolean {
        if (num < 2) return false
        
        for (i in 2..sqrt(num.toDouble()).toInt()) {
            if (num % i == 0) return false
        }
        
        return true
    }
    
    fun solution(numbers: String): Int {
        val answer = makePermutation(numbers.toList())
        
        return answer.count {isPrime(it)}
    }
}